### PR TITLE
Add additional unit and integration tests for etag functionality

### PR DESCRIPTION
**Pull request by sandhyasun**
**Description**.
- Unit tests for corner cases  and interaction with the IcebergHttpUtil etag generation method 
- Unit test for  etag changes in the context of schema lifecycle changes 
- Integration tests  for  non-existent tables with etag passed in by client, malformed headers
- Integration tests for DDL and DML changes affecting etags

**Motivation.**
The additonal tests ensure that the ETag mechanism correctly handles the most common table changes -  schema evolution and DDL/ DML operations. This ensures caching correctness in scenarios where tables are actively being written to and callers are caching etags for performance reasons

**Related change**
Related original change [(Commit 8b5dfa9)](https://github.com/apache/polaris/commit/8b5dfa97a8e50a53412d2ded40f2e4db5a40cc42)

### CODE REVIEW

File: service/common/src/test/java/org/apache/polaris/service/http/IfNoneMatchTest.java [hunk 1]
```diff
@@ -129,4 +136,193 @@ public void invalidHeaderWithOnlyWhitespacesBetween() {
...
+
+    String originalETag = IcebergHttpUtil.generateETagForMetadataFileLocation(originalMetadataLocation);
+    String updatedETag = IcebergHttpUtil.generateETagForMetadataFileLocation(updatedMetadataLocation);
+
+    // ETags should be different for different metadata locations
+    assertThat(originalETag).isNotEqualTo(updatedETag);
+
+    // Both should be valid weak ETags
+    assertThat(originalETag).startsWith("W/\"").endsWith("\"");
+    assertThat(updatedETag).startsWith("W/\"").endsWith("\"");
```
Comment (dimas-b) [inline service/common/src/test/java/org/apache/polaris/service/http/IfNoneMatchTest.java:170]: nit: the lines above appear to subsume `testETagGenerationDifferentLocations`

File: service/common/src/test/java/org/apache/polaris/service/http/IfNoneMatchTest.java [hunk 2]
```diff
@@ -129,4 +136,193 @@ public void invalidHeaderWithOnlyWhitespacesBetween() {
...
+    // Should match v1 (one of the ETags in the list)
+    assertThat(multipleIfNoneMatch.anyMatch(v1ETag)).isTrue();
+
+    // Should NOT match v2 or v3 (not in the list)
+    assertThat(multipleIfNoneMatch.anyMatch(v2ETag)).isFalse();
+    assertThat(multipleIfNoneMatch.anyMatch(v3ETag)).isFalse();
+  }
+
+  @Test
+  public void testETagBehaviorForTableDropAndRecreate() {
```
Comment (dimas-b) [inline service/common/src/test/java/org/apache/polaris/service/http/IfNoneMatchTest.java]: Is this test materially different from `testETagChangeAfterMetadataLocationChange`?

File: service/common/src/test/java/org/apache/polaris/service/http/IfNoneMatchTest.java [hunk 3]
```diff
@@ -129,4 +136,193 @@ public void invalidHeaderWithOnlyWhitespacesBetween() {
...
+    String recreatedV2MetadataLocation = baseLocation + "recreated-v2.metadata.json";
+    String recreatedV2ETag = IcebergHttpUtil.generateETagForMetadataFileLocation(recreatedV2MetadataLocation);
+
+    // All ETags should be unique
+    List<String> allETags = List.of(v1ETag, v2ETag, v3ETag, recreatedV1ETag, recreatedV2ETag);
+
+    // Verify all ETags are different from each other
+    for (int i = 0; i < allETags.size(); i++) {
+      for (int j = i + 1; j < allETags.size(); j++) {
+        assertThat(allETags.get(i)).isNotEqualTo(allETags.get(j));
```
Comment (dimas-b) [inline service/common/src/test/java/org/apache/polaris/service/http/IfNoneMatchTest.java:260]: nit: might be more intuitive to put them all into a `Set` and check size.

File: integration-tests/src/main/java/org/apache/polaris/service/it/test/PolarisRestCatalogIntegrationTest.java [hunk 4]
```diff
@@ -1549,4 +1554,412 @@ public void testUpdateTableWithReservedProperty() {
...
+    // Test with current v3 ETag
+    Invocation currentEtagInvocation =
+        catalogApi
+            .request("v1/" + currentCatalogName + "/namespaces/ns1/tables/test_schema_evolution_table")
+            .header(HttpHeaders.IF_NONE_MATCH, v3ETag)
+            .build("GET");
+
+    try (Response currentEtagResponse = currentEtagInvocation.invoke()) {
+      // Should return 304 Not Modified because ETag matches current version
+      assertThat(currentEtagResponse.getStatus()).isEqualTo(Response.Status.NOT_MODIFIED.getStatusCode());
```
Comment (dimas-b) [inline integration-tests/src/main/java/org/apache/polaris/service/it/test/PolarisRestCatalogIntegrationTest.java]: Would it also make sense to test this when `v3ETag` is combined with `v1ETag`?

File: integration-tests/src/main/java/org/apache/polaris/service/it/test/PolarisRestCatalogIntegrationTest.java [hunk 5]
```diff
@@ -1549,4 +1554,412 @@ public void testUpdateTableWithReservedProperty() {
...
+
+    // Simulate DML operation by creating a new snapshot (append operation)
+    Table table = restCatalog.loadTable(tableId);
+
+    // Create a data file and append it (simulating INSERT operation)
+    AppendFiles append = table.newAppend();
+
+    // Create a mock data file entry
+    DataFile dataFile = DataFiles.builder(table.spec())
+        .withPath("/path/to/data/file1.parquet")
```
Comment (dimas-b) [inline integration-tests/src/main/java/org/apache/polaris/service/it/test/PolarisRestCatalogIntegrationTest.java]: not sure this will work with cloud storage :thinking: Why not `table.locationProvider().newDataLocation(...)`?

**Review by dimas-b — COMMENTED** [review]
Thanks for your contribution, @sandhyasun !

The tests looks very useful to me overall... just some minor comments.

File: service/common/src/test/java/org/apache/polaris/service/http/IfNoneMatchTest.java (hunk 1 above)
Comment (sandhyasun) [reply service/common/src/test/java/org/apache/polaris/service/http/IfNoneMatchTest.java:170]: Agreed - consolidated 3 tests into 1 that basically test metadata location change.

File: service/common/src/test/java/org/apache/polaris/service/http/IfNoneMatchTest.java (hunk 2 above)
Comment (sandhyasun) [reply service/common/src/test/java/org/apache/polaris/service/http/IfNoneMatchTest.java]: Agreed - consolidated 3 tests into 1 that basically test metadata location change.

File: service/common/src/test/java/org/apache/polaris/service/http/IfNoneMatchTest.java (hunk 3 above)
Comment (sandhyasun) [reply service/common/src/test/java/org/apache/polaris/service/http/IfNoneMatchTest.java:260]: Done

File: integration-tests/src/main/java/org/apache/polaris/service/it/test/PolarisRestCatalogIntegrationTest.java (hunk 4 above)
Comment (sandhyasun) [reply integration-tests/src/main/java/org/apache/polaris/service/it/test/PolarisRestCatalogIntegrationTest.java]: Added another case

File: integration-tests/src/main/java/org/apache/polaris/service/it/test/PolarisRestCatalogIntegrationTest.java (hunk 5 above)
Comment (sandhyasun) [reply integration-tests/src/main/java/org/apache/polaris/service/it/test/PolarisRestCatalogIntegrationTest.java]: True ! Fixed it.

File: service/common/src/test/java/org/apache/polaris/service/http/IfNoneMatchTest.java (hunk 3 above)
Comment (dimas-b) [reply service/common/src/test/java/org/apache/polaris/service/http/IfNoneMatchTest.java:260]: Did you forget to push? :sweat_smile:

Comment (sandhyasun) [reply service/common/src/test/java/org/apache/polaris/service/http/IfNoneMatchTest.java:260]: This 4th commit should show the change ?  [https://github.com/apache/polaris/pull/1972/commits/770b6bce05803f4bec45e61b38f437050986b023](url)

Comment (dimas-b) [reply service/common/src/test/java/org/apache/polaris/service/http/IfNoneMatchTest.java:260]: I still see nested loops :shrug: but it's a nit comment... approving.
